### DISCUSSION TITLE
LSP and Python: flake8 settings are not honored

### DISCUSSION THREAD

**Discussion by synic** [syl20bnr/spacemacs#15138] [Q&A]
#### Description :octocat:
LSP and Python: flake8 settings are not honored

#### Reproduction guide :beetle:
- Start Emacs
- Create a new directory, something like `pythonlsptest`
- In the directory, type `git init`
- Create a `.flake8` configuration with the following contents:

    ```ini
    [flake8]
    ignore = D100,D101,D102,D103,D104,D105,D106,D107,W503,E203
    application-import-names = pythonlsptest
    max-line-length = 88
    import-order-style = google
    ```
- Create a python file with a function in it:

    ```python

    def test():
        print('test')

    ```
    
*Observed behaviour:* :eyes: :broken_heart:

Emacs ignores the flake8 configuration. Errors that are set to be ignored
(for example, `D100`, `D103`) are shown in flyckeck.

I have tried telling LSP where the configuration file is using `.dir-locals.el`:

```lisp
((python-mode
  (eval . (python-black-on-save-mode))
  (fill-column . 88)
  (auto-fill-mode . -1)
  (flycheck-flake8rc . "/pythonlsptest/.flake8")
  (lsp-pylsp-plugins-flake8-config . "/pythonlsptest/.flake8")
 ))
```

However, this appears to have no effect.

*Expected behaviour:* :heart: :smile:

The flake8 configuration should be honored like it is in `company-mode`

#### System Info :computer:
- OS: gnu/linux
- Emacs: 27.1
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 500335a2c)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(auto-completion colors csv dap django emacs-lisp git go helm html ipython-notebook javascript lua markdown major-modes nginx org osx perl5 python rust spacemacs-layouts spell-checking swift syntax-checking sql yaml vimscript
                 (c-c++ :variables c-c++-enable-clang-support t)
                 (lsp :variables lsp-headerline-breadcrumb-enable nil)
                 (shell :variables shell-default-height 30 shell-default-position 'bottom)
                 (typescript :variables typescript-backend 'lsp typescript-fmt-tool 'prettier typescript-fmt-on-save t)
                 (version-control :variables version-control-diff-tool 'diff-hl))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE HARFBUZZ M17N_FLT LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP

#### Backtrace :paw_prints:
```
<<BACKTRACE IF RELEVANT>>
```

**Comment by lebensterben** [comment]
try M-x flycheck-select-checker and choose python-flake8

**Comment by synic** [comment]
@lebensterben This does work, thanks! How do I get it to automatically select `python-flake8`?

I've got this setup in my `.dir-locals.el`: 

```lisp
((python-mode
  (eval . (python-black-on-save-mode))
  (fill-column . 88)
  (auto-fill-mode . -1)
  (flycheck--automatically-disabled-checkers . '(python-pylint lsp python-pycompile))
  (flycheck--automatically-enabled-checkers . '(python-flake8))
  (flycheck-add-next-checker 'python-flake8 'lsp)
))

```

but it sure doesn't want to disable lsp as a checker

**Comment by lebensterben** [comment]
set flycheck-checker

**Comment by synic** [comment]
It doesn't work, for some reason lsp gets enabled by default, even with the following:

```lisp
((python-mode
  (eval . (python-black-on-save-mode))
  (fill-column . 88)
  (auto-fill-mode . -1)
  (flycheck-flake8rc . "/home/synic/Projects/defiq-coinprice-lambda/.flake8")
  (lsp-pylsp-plugins-flake8-config . "/home/synic/Projects/defiq-coinprice-lambda/.flake8")
  (flycheck-checker . 'python-flake8)
))
```

I've got this in my spacemacs config as well:

```lisp
  (add-hook 'python-mode-hook
            (lambda ()
              ;; Enable fill column indicator
              ;; (display-fill-column-indicator-mode)
              (display-fill-column-indicator-mode)
              ;; Turn on line numbering
              ;; (linum-mode t)
              (setq fill-column 79)
              ;; set in-block indentation scale
              (setq python-indent-def-block-scale 1)
              ;; Set tab-width to 4
              (setq tab-width 4
                    evil-shift-width 4)
              (setq flycheck-checker 'python-flake8)
              ;; Enable automatic line wrapping at fill column
              (auto-fill-mode t)))

```

When I check the flycheck setup, is shows lsp as "explicitly selected"

**Comment by lebensterben** [comment]
lsp blackend is enabled whenever lsp layer is enabled. you have to explicitly set python layer variable to disable it.

this is not a bug and I'm transferring it to discussion.

**Reply by synic** [reply]
How do you do that? I don't mind LSP and it works fine after setting the checker to `python-flake8`, but there doesn't seem to be a way to set it up in configuration. I have to open the file and then run `(setq flycheck-checker 'python-flake8)`

**Reply by swinkels** [reply]
Hope I understood your question correctly: if you set variable `lsp-diagnostics-provider` to `:none`, lsp-mode will not replace the standard flycheck checkers.

I used the Customize interface to set that value: I selected `4 = Use neither flymake nor lsp`, saved the settings for future sessions and restarted Emacs. After the restart, the flycheck buffer was populated using the standard Python flycheck checkers, e.g.
```
 File    Line Col Level    ID     Message (Checker)
 fabfil... 70  27 error           "DeployEvent" has no attribute "path_to_workspace" (python-mypy)
 fabfil... 71  35 error           "DeployEvent" has no attribute "path_to_virtualenv" (python-mypy)
```
Note the `(python-mypy)` instead of `(lsp)` as providing the input. 

At the moment I prefer to use the standard Python flycheck checkers as issues with them are much to debug than issues in an lsp server. But maybe I just need to get a bit of confidence in the lsp server - by the way, I use pylsp.

Here's the output to my `flycheck-verify-setup`:
```
Syntax checkers for buffer fabfile.py in python-mode:

First checker to run:

  python-flake8
    - may enable:         yes
    - executable:         Found at /home/zzfrd/repos/git/my-app/.tox/py38-dev/bin/python3
    - configuration file: Not found
    - `flake8' module:    Found at "/home/zzfrd/repos/git/my-app/.tox/py38-dev/lib64/python3.8/site-packages/flake8/__init__.py"
    - next checkers:      python-pylint, python-mypy

Checkers that may run as part of the first checker's chain:

  python-mypy
    - may enable:         yes
    - may run:            t
    - executable:         Found at /home/zzfrd/repos/git/my-app/.tox/py38-dev/bin/mypy
    - configuration file: Found at "/home/zzfrd/repos/git/my-app/mypy.ini"

Checkers that could run if selected:

# etc...
```

**Reply by lebensterben** [reply]
For choosing lsp vs. non-lsp backend, read python layer documentation.

Also see @swinkels 's comment for a more detailed workaround.
